File: client/src/services/vacancy.js
import axios from '../utils/axios';


export async function createNewVacancy(name, description, place, salary, experience, tags, setError, setShowError, navigate) {
    try {
        const response = await axios.post('/vacancies/create', { name, description, place, salary, experience, tags });

        if (response.data) {
            navigate('/employers');
            window.location.reload(); 
            console.log(response.data);
        }
    } catch (error) {
        const errorMessage = error.response.data.detail;
        setError(errorMessage);
        setShowError(true);
        setTimeout(() => {
          setShowError(false);
          setError(null);
        }, 2500);
      }    
}


export async function getVacanciesByEmployer(employer_id) {
    try {
        const response = await axios.get(`/vacancies/employer/${employer_id}`);

        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function getFilteredVacancies(place, salary, experience) {
    try {
        const response = await axios.get(`/vacancies/filter?place=${place}&salary=${salary}&experience=${experience}`);
        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}



export async function getPaginatedVacancies(current_page, confirmed, archived) {
    try {
        const response = await axios.get(`/vacancies/?page=${current_page}&confirmed=${confirmed}&archived=${archived}`);

        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function searchVacancies(query) {
    try {
      const response = await axios.get(`/vacancies/search?query=${query}`);
  
      if (response.data) {
        return response.data;
      }
    } catch (error) {
      console.log(error.response.data.detail);
    }
}  


export async function deleteVacancy(vacancy_id, vacancies, setVacancies) {
    try {
        const response = await axios.delete(`/vacancies/delete/${vacancy_id}`);

        if (response.data) {
            setVacancies(vacancies.filter(vacancy => vacancy.id !== vacancy_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function confirmVacancy(vacancy_id, vacancies, setVacancies) {
    try {
        const response = await axios.patch(`/vacancies/confirm/${vacancy_id}`);

        if (response.data) {
            setVacancies(vacancies.filter(vacancy => vacancy.id !== vacancy_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function inArchiveVacancy(vacancy_id, vacancies, setVacancies) {
    try {
      const response = await axios.patch(`/vacancies/in_archive/${vacancy_id}`);
  
      if (response.data) {
        setVacancies(vacancies.filter(vacancy => vacancy.id !== vacancy_id));
        console.log(response.data);
      }
    } catch (error) {
      console.log(error.response.data.detail);
    }
  }
  

export async function fromArchiveVacancy(vacancy_id, vacancies, setVacancies) {
    try {
        const response = await axios.patch(`/vacancies/from_archive/${vacancy_id}`);

        if (response.data) {
            setVacancies(vacancies.filter(vacancy => vacancy.id !== vacancy_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}

File: server/routes/vacancy.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from config.database import get_db

from dto.vacancy import Vacancy as VacancyDTO
from services.user import get_user_id
from services import vacancy as VacancyService


router = APIRouter()


@router.post("/vacancies/create")
async def create_vacancy(data: VacancyDTO, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.create_vacancy(data, user_id, db)


@router.get("/vacancies/employer/{employer_id}")
async def get_vacancies_by_employer_id(employer_id: str, db: Session = Depends(get_db)):
    return VacancyService.get_vacancies_by_employer_id(employer_id, db)


@router.get("/vacancies")
async def get_paginated_vacancies(page: int, confirmed: bool, archived: bool, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.get_paginated_vacancies(page, confirmed, archived, user_id, db)


@router.get("/vacancies/search")
async def search_vacancies(query: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.search_vacancies(query, user_id, db)


@router.patch("/vacancies/confirm/{vacancy_id}")
async def confirm_vacancy(vacancy_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.confirm_vacancy(vacancy_id, user_id, db)


@router.patch("/vacancies/in_archive/{vacancy_id}")
async def in_archive_vacancy(vacancy_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.in_archive_vacancy(vacancy_id, user_id, db)


@router.patch("/vacancies/from_archive/{vacancy_id}")
async def from_archive_vacancy(vacancy_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.from_archive_vacancy(vacancy_id, user_id, db)


@router.delete("/vacancies/delete/{vacancy_id}")
async def delete_vacancy(vacancy_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return VacancyService.delete_vacancy(vacancy_id, user_id, db)
